### ISSUE TITLE
yum module does not work with custom RPMs

### ISSUE THREAD

**Issue by ciupicri** [ansible/ansible#1754]
The yum module does not work properly when using custom RPMs with versions different than the ones available in the yum repositories.

Software used:
- management machine (hermes): ansible 637883164edebb911b185e53b76fdfba14fa7d41 and Fedora 17 x86_64
- orchestrated machine (vm): Fedora 18 x86_64 Beta

Here is what is happening when managing the python-trml2pdf RPM.
1.  python-trml2pdf is not installed, so we'll install by hand an old version of it:
   
   ```
   [root@vm ~]# yum install ~/rpmbuild/RPMS/noarch/python-trml2pdf-1.2-2.fc18.noarch.rpm 
   ```
2. Next, use ansible to make sure it's installed:
   
   ```
   [ciupicri@hermes ~]$ ansible vms  -m yum -a "pkg=python-trml2pdf state=installed"
   192.168.122.219 | success >> {
       "changed": true, 
       "msg": "", 
       "rc": 0, 
       "results": [
           "\n================================================================================\n Package                 Arch           Version            Repository      Size\n================================================================================\nUpdating:\n python-trml2pdf         noarch         1.2-5.fc18         fedora          33 k\n\nTransaction Summary\n================================================================================\nUpgrade  1 Package\n\nTotal size: 33 k\n\nUpdated:\n  python-trml2pdf.noarch 0:1.2-5.fc18                                           \n\n"
       ]
   }
   ```
   
   Notice that **changed** is true, though it shouldn't.
3.  Check what happened on the orchestrated machine:
   
   ```
   [root@vm ~]# rpm -q python-trml2pdf
   python-trml2pdf-1.2-5.fc18.noarch
   ```
   
   python-trml2pdf was upgraded to 1.2-5.fc18.noarch (the latest version available in the yum repositories) even if we didn't request this.
4. Re-run ansible:
   
   ```
   [ciupicri@hermes ~]$ ansible vms  -m yum -a "pkg=python-trml2pdf state=installed"
   192.168.122.219 | success >> {
       "changed": false, 
       "msg": "", 
       "rc": 0, 
       "results": [
           "python-trml2pdf-1.2-5.fc18.noarch providing python-trml2pdf is already installed"
       ]
   }
   ```
   
   **changed** is false just as it should be.
5.  Install by hand a newer version python-trml2pdf:
   
   ```
   [root@vm ~]# yum install ~/rpmbuild/RPMS/noarch/python-trml2pdf-1.2-9.fc18.noarch.rpm 
   ```
6.  Re-run ansible:
   
   ```
   [ciupicri@hermes ~]$ ansible vms  -m yum -a "pkg=python-trml2pdf state=installed"
   192.168.122.219 | success >> {
       "changed": true, 
       "msg": "", 
       "rc": 0, 
       "results": [
           "Package matching python-trml2pdf-1.2-5.fc18.noarch already installed. Checking for update.\n"
       ]
   }
   ```
   
   **changed** is true even if the package was already installed.
7. Check what happened on the orchestrated machine:
   
   ```
   [root@vm ~]# rpm -q python-trml2pdf
   python-trml2pdf-1.2-9.fc18.noarch
   ```

**Comment by skvidal** [comment]
when you have one of the lower-version local installed versions installed please run this command and report the output

repoquery --disablerepo=\*  --pkgnarrow=installed  python-trml2pdf

and run

repoquery  python-trml2pdf

thanks

**Comment by ciupicri** [comment]
```
[root@vm ~]# yum install ~/rpmbuild/RPMS/noarch/python-trml2pdf-1.2-2.fc18.noarch.rpm 
...
Installed:
  python-trml2pdf.noarch 0:1.2-2.fc18                                               

Complete!
[root@vm ~]# repoquery --disablerepo=* --pkgnarrow=installed python-trml2pdf
python-trml2pdf-0:1.2-2.fc18.noarch
[root@vm ~]# repoquery python-trml2pdf
python-trml2pdf-0:1.2-5.fc18.noarch
```

**Comment by skvidal** [comment]
Can you apply a patch and test it for me?

**Comment by skvidal** [comment]
http://paste.fedoraproject.org/2361/

**Comment by ciupicri** [comment]
I start with the oldest version available:

```
[root@vm ~]# rpm -q python-trml2pdf
python-trml2pdf-1.2-2.fc18.noarch
```

ansible seems to work:

```
[ciupicri@hermes ~]$ ansible vms  -m yum -a "pkg=python-trml2pdf state=installed"
192.168.122.219 | success >> {
    "changed": false, 
    "msg": "", 
    "rc": 0, 
    "results": [
    "package providing python-trml2pdf is already installed"
    ]
}
```

I update manually to the intermediate version:

```
[root@vm ~]# yum -y update python-trml2pdf
...
Updated:
  python-trml2pdf.noarch 0:1.2-5.fc18
```

ansible works correctly again:

```
[ciupicri@hermes ~]$ ansible vms  -m yum -a "pkg=python-trml2pdf state=installed"
192.168.122.219 | success >> {
    "changed": false, 
    "msg": "", 
    "rc": 0, 
    "results": [
    "python-trml2pdf-1.2-5.fc18.noarch providing python-trml2pdf is already installed"
    ]
}
```

I upgrade manually to the latest version:

```
[root@vm ~]# yum install ~/rpmbuild/RPMS/noarch/python-trml2pdf-1.2-9.fc18.noarch.rpm 
...
Updated:
  python-trml2pdf.noarch 0:1.2-9.fc18
```

ansible saves the day:

```
[ciupicri@hermes ~]$ ansible vms  -m yum -a "pkg=python-trml2pdf state=installed"
192.168.122.219 | success >> {
    "changed": false, 
    "msg": "", 
    "rc": 0, 
    "results": [
    "package providing python-trml2pdf is already installed"
    ]
}
```

**tl;dr** the patch seems to work

**Comment by skvidal** [comment]
https://github.com/ansible/ansible/pull/1760

**Comment by dhozac** [comment]
Closing thanks to #1760.
